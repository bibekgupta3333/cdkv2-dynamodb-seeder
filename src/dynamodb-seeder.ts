import * as path from "path";
import {
  CustomResource,
  Duration,
  aws_iam as iam,
  aws_dynamodb as dynamodb,
  aws_s3 as s3
} from "aws-cdk-lib";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

import { Seeds } from "./seeds";

export interface DynamoDBSeederProps {
  readonly table: dynamodb.ITable;
  readonly seeds: Seeds;

  /**
   * The function execution time (in seconds) after which Lambda terminates
   * the function. Because the execution time affects cost, set this value
   * based on the function's expected execution time.
   *
   * @default Duration.minutes(15)
   */
  readonly timeout?: Duration;
}

export class DynamoDBSeeder extends Construct {
  constructor(scope: Construct, id: string, props: DynamoDBSeederProps) {
    super(scope, id);

    const seeds = props.seeds.bind(this);
    const seedsBucket = seeds.s3Location?.bucketName
      ? s3.Bucket.fromBucketName(this, "SeedsBucket", seeds.s3Location.bucketName)
      : undefined;
    const entryPath = path.join(__dirname, "lambdas", "dynamodb-seeder", "index.js");
    console.log(entryPath);
    const handler = new NodejsFunction(this, "CustomResourceHandler", {
      runtime: Runtime.NODEJS_16_X,
      entry: path.join(__dirname, "lambdas", "dynamodb-seeder", "index.js"),
      handler: "handler",
      memorySize: 512,
      timeout: props.timeout ?? Duration.minutes(15),
      bundling: {
        minify: true, // minify code, defaults to false
        sourceMap: true,
        target: "es2020", // target environment for the generated JavaScript code
        metafile: true // include meta file, defaults to false
      },
      environment: {
        NODE_OPTIONS: "--enable-source-maps"
      }
    });

    handler.addToRolePolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["dynamodb:BatchWriteItem"],
        resources: [props.table.tableArn]
      })
    );

    if (props.table.encryptionKey) {
      handler.addToRolePolicy(
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            "kms:Encrypt",
            "kms:Decrypt",
            "kms:ReEncrypt*",
            "kms:GenerateDataKey*",
            "kms:DescribeKey",
            "kms:CreateGrant"
          ],
          resources: [props.table.encryptionKey.keyArn]
        })
      );
    }

    if (seedsBucket) {
      const objectKey = seeds.s3Location?.objectKey ?? "*";

      handler.addToRolePolicy(
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ["s3:GetObject"],
          resources: [seedsBucket.arnForObjects(objectKey)]
        })
      );
    }

    new CustomResource(this, "CustomResource", {
      serviceToken: handler.functionArn,
      resourceType: "Custom::DynamodbSeeder",
      properties: {
        TableName: props.table.tableName,
        Seeds: {
          InlineSeeds: seeds.inlineSeeds,
          S3Bucket: seeds.s3Location && seeds.s3Location.bucketName,
          S3Key: seeds.s3Location && seeds.s3Location.objectKey,
          S3ObjectVersion: seeds.s3Location && seeds.s3Location.objectVersion
        }
      }
    });
  }
}
