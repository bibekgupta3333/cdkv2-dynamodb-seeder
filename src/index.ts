export * from "./dynamodb-seeder";
export * from "./seeds";
