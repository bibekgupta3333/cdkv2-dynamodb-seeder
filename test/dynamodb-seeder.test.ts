import * as path from "path";
import { Stack } from "aws-cdk-lib";
import { AttributeType, Table, TableEncryption } from "aws-cdk-lib/aws-dynamodb";
import { Bucket } from "aws-cdk-lib/aws-s3";
import "jest-cdk-snapshot";

import { DynamoDBSeeder } from "../src/dynamodb-seeder";
import { Seeds } from "../src/seeds";

test("inline", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    }
  });

  // WHEN
  new DynamoDBSeeder(stack, "DynamoDBSeeder", {
    table,
    seeds: Seeds.fromInline([
      {
        id: 1,
        column: "foo"
      },
      {
        id: 2,
        column: "bar"
      }
    ])
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});

test("json file", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    }
  });

  // WHEN
  new DynamoDBSeeder(stack, "DynamoDBSeeder", {
    table,
    seeds: Seeds.fromJsonFile(path.join(__dirname, "mocks", "seeds.json"))
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});

test("json file: no such file", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    }
  });

  // WHEN
  const addSeeder = () => {
    new DynamoDBSeeder(stack, "DynamoDBSeeder", {
      table,
      seeds: Seeds.fromJsonFile(path.join(__dirname, "mocks", "XYZ.json"))
    });
  };

  // THEN
  expect(() => addSeeder()).toThrowError(/Cannot find asset at/);
});

test("json file: no json file", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    }
  });

  // WHEN
  const addSeeder = () => {
    new DynamoDBSeeder(stack, "DynamoDBSeeder", {
      table,
      seeds: Seeds.fromJsonFile(path.join(__dirname, "mocks", "nojson.txt"))
    });
  };

  // THEN
  expect(() => addSeeder()).toThrowError(new Error("Could not convert file to JSON"));
});

test("bucket", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    }
  });

  const seedsBucket = Bucket.fromBucketName(stack, "SeedsBucket", "my-seeds-bucket");

  // WHEN
  new DynamoDBSeeder(stack, "DynamoDBSeeder", {
    table,
    seeds: Seeds.fromBucket(seedsBucket, "seeds.json")
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});

test("multiple seeders", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    }
  });

  const seedsBucket1 = Bucket.fromBucketName(stack, "SeedsBucket1", "seeds-bucket-one");
  const seedsBucket2 = Bucket.fromBucketName(stack, "SeedsBucket2", "seeds-bucket-two");

  // WHEN
  new DynamoDBSeeder(stack, "DynamoDBSeeder1", {
    table,
    seeds: Seeds.fromBucket(seedsBucket1, "seeds-one.json")
  });
  // WHEN
  new DynamoDBSeeder(stack, "DynamoDBSeeder2", {
    table,
    seeds: Seeds.fromBucket(seedsBucket2, "seeds-two.json")
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});

test("customer managed encryption key", () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, "Table", {
    partitionKey: {
      name: "id",
      type: AttributeType.NUMBER
    },
    encryption: TableEncryption.CUSTOMER_MANAGED
  });

  // WHEN
  new DynamoDBSeeder(stack, "DynamoDBSeeder", {
    table,
    seeds: Seeds.fromInline([
      {
        id: 1,
        column: "foo"
      },
      {
        id: 2,
        column: "bar"
      }
    ])
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});
