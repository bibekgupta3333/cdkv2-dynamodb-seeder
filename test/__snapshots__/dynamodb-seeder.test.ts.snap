// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bucket 1`] = `
{
  "Resources": {
    "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10": {
      "DependsOn": [
        "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0cca44bd7835a9838d695ed9cbc4da0e36d7d616850d92d576098b96046e56e4.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableCD117FA1",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":s3:::my-seeds-bucket/seeds.json",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "Roles": [
          {
            "Ref": "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DynamoDBSeederCustomResourceEF4D5C60": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Seeds": {
          "S3Bucket": "my-seeds-bucket",
          "S3Key": "seeds.json",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10",
            "Arn",
          ],
        },
        "TableName": {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`customer managed encryption key 1`] = `
{
  "Resources": {
    "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10": {
      "DependsOn": [
        "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0cca44bd7835a9838d695ed9cbc4da0e36d7d616850d92d576098b96046e56e4.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableCD117FA1",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kms:Encrypt",
                "kms:Decrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
                "kms:DescribeKey",
                "kms:CreateGrant",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableKey25666F95",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "Roles": [
          {
            "Ref": "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DynamoDBSeederCustomResourceEF4D5C60": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Seeds": {
          "InlineSeeds": "[{"id":1,"column":"foo"},{"id":2,"column":"bar"}]",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10",
            "Arn",
          ],
        },
        "TableName": {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "SSESpecification": {
          "KMSMasterKeyId": {
            "Fn::GetAtt": [
              "TableKey25666F95",
              "Arn",
            ],
          },
          "SSEEnabled": true,
          "SSEType": "KMS",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TableKey25666F95": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "Customer-managed key auto-created for encrypting DynamoDB table at Default/Table",
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`inline 1`] = `
{
  "Resources": {
    "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10": {
      "DependsOn": [
        "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0cca44bd7835a9838d695ed9cbc4da0e36d7d616850d92d576098b96046e56e4.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableCD117FA1",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "Roles": [
          {
            "Ref": "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DynamoDBSeederCustomResourceEF4D5C60": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Seeds": {
          "InlineSeeds": "[{"id":1,"column":"foo"},{"id":2,"column":"bar"}]",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10",
            "Arn",
          ],
        },
        "TableName": {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`json file 1`] = `
{
  "Resources": {
    "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10": {
      "DependsOn": [
        "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0cca44bd7835a9838d695ed9cbc4da0e36d7d616850d92d576098b96046e56e4.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableCD117FA1",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":s3:::",
                    {
                      "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                    },
                    "/d030fc4499f9b92060beabbe94d500033cde3dd8a61a1668fdfcee96b2019ffd.json",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "Roles": [
          {
            "Ref": "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DynamoDBSeederCustomResourceEF4D5C60": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Seeds": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "d030fc4499f9b92060beabbe94d500033cde3dd8a61a1668fdfcee96b2019ffd.json",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10",
            "Arn",
          ],
        },
        "TableName": {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`multiple seeders 1`] = `
{
  "Resources": {
    "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10": {
      "DependsOn": [
        "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0cca44bd7835a9838d695ed9cbc4da0e36d7d616850d92d576098b96046e56e4.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableCD117FA1",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":s3:::seeds-bucket-one/seeds-one.json",
                  ],
                ],
              },
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":s3:::seeds-bucket-two/seeds-two.json",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomDynamodbSeederCustomDynamodbSeederServiceRoleDefaultPolicy6C04255C",
        "Roles": [
          {
            "Ref": "CustomDynamodbSeederCustomDynamodbSeederServiceRole5FC6AEF8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DynamoDBSeeder1CustomResource555E70AD": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Seeds": {
          "S3Bucket": "seeds-bucket-one",
          "S3Key": "seeds-one.json",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10",
            "Arn",
          ],
        },
        "TableName": {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "DynamoDBSeeder2CustomResourceAC955965": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Seeds": {
          "S3Bucket": "seeds-bucket-two",
          "S3Key": "seeds-two.json",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDynamodbSeederCustomDynamodbSeeder3D3D7F10",
            "Arn",
          ],
        },
        "TableName": {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
